### ISSUE TITLE
Shader errors on intel HD graphics

### ISSUE THREAD

**Issue by germn** [tobspr/RenderPipeline#54]
Trying to run any of example leads to error on intel HD videocard:

> [>] ScatteringMethodEricBruneton Precomputing ...
> :display:gsg:glgsg(error): SHADER_ID_COMPILE error has been generated. GLSL compile failed for shader 2, "/$$rp/rpplugins/scattering/shader/eric_bruneton/transmittance.compute.glsl": ERROR: 0:34: 'restrict' : syntax error syntax error

I tried to remove all "RESTRICT" keywords from shaders. Looks like shaders compiles, but I got warnings:

> :display:gsg:glgsg(warning): SHADER_ID_COMPILE other warning has been generated. GLSL compile warning(s) for shader 18, "/$$rp/rpplugins/scattering/shader/eric_bruneton/add_delta_e.compute.glsl": WARNING: 0:38: 'image format' : redundant image format layout qualifier

[...]

> :display:gsg:glgsg(warning): SHADER_ID_COMPILE other warning has been generated. GLSL compile warning(s) for shader 20, "/$$rp/rpplugins/scattering/shader/eric_bruneton/add_delta_sr.compute.glsl": WARNING: 0:38: 'image format' : redundant image format layout qualifier

And after this warning interpreter crashes.

Windows 10
Intel HD 4600
Latest Panda dev build and Pipeline

**Comment by tobspr** [comment]
Weird, I will check in a fix for the restrict stuff. For the layout qualifiers, I am not sure if thats not a buggy driver - do you have the latest drivers?

**Comment by germn** [comment]
Yes, latest driver from Intel. I think I'll try to write them about this warning.

Does warning (not error) itself lead interpreter crash?

**Comment by tobspr** [comment]
Well the interpreter crash is probably caused by the driver crashing. It looks to me like there is a bug in the driver - restrict is a valid keyword according to the OpenGL standard, and the layout() qualifier is also correct.

**Comment by germn** [comment]
Thanks for help, reported all to Intel:
https://software.intel.com/en-us/forums/graphics-driver-bug-reporting/topic/625013
https://software.intel.com/en-us/forums/graphics-driver-bug-reporting/topic/625014

**Comment by tobspr** [comment]
RESTRICT is just a define by the render pipeline. it should be lowercase - "restrict", not "RESTRICT". Otherwise the support will probably thing its an issue of lower/uppercase.

**Comment by germn** [comment]
Thanks again, fixed it.

**Comment by tobspr** [comment]
Actually, can you try the following for the layout bug:

layout(rgba16f) uniform image3D xxx;

Maybe intel expects it to be in a different order. I.e. like nvidia cards don't support restrict (their compiler crashes on that keywoard), and amd expects it to be in a different position than the spec says.

**Comment by germn** [comment]
You're right! After putting layout at start, warning disappear.
Looks like pipeline loads, but interpreter still crashes :)

```
[>] CORE                      Using native core module
[>] RenderPipeline            Using Python 2.7 with architecture win_i386
[>] RenderPipeline            Using Panda3D 1.10.0 built on Apr 17 2016 14:46:39
[>] RenderPipeline            Using git commit 380ff447101e6809d90c9edec7781bbaa339d894
[>] MountManager              Auto-Detected base path to /c/Users/gmn/Dropbox/gmn/projects/_test/Python/Panda3D/RenderPipeline
[>] RenderPipeline            Mount manager was not mounted, mounting now ...
[>] MountManager              Setting up virtual filesystem
[>] MountManager              Mounting auto-detected config dir: /c/Users/gmn/Dropbox/gmn/projects/_test/Python/Panda3D/RenderPipeline/config/
[>] MountManager              Mounting ramdisk as /$$rptemp
[>] RenderPipeline            No settings loaded, loading from default location 
Known pipe types:
  wglGraphicsPipe
(all display modules loaded.)
:display:gsg:glgsg(warning): API_ID_DEPRECATED_HINT deprecated behavior warning has been generated. Using of GL_FOG_HINT hint is deprecated.
[>] RenderPipeline            Render resolution is 800 x 600
[>] LightManager              Tile size = 24 x 16 , Num tiles = 34 x 38
[>] GPUCommandQueue           Allocating command buffer of size 4096
[>] GroupedInputBlock         Native UBO support = False
[>] PluginManager             Loading plugin settings
[>] PluginManager             Creating plugin instances ..
[>] Debugger                  Creating debugger
[>] plugin:scattering         Loading scattering method for 'eric_bruneton'
[>] StageManager              Setup stages ..
[>] StageManager              Preparing stages ..
[>] CollectUsedCellsStage     Allocating 41344 cells
[>] CullLightsStage           Using 162 culling lines
[>] StageManager              Storing previous frame pipe for AmbientOcclusion
[>] StageManager              Storing previous frame pipe for SceneDepth
[>] StageManager              Storing previous frame pipe for PostAmbientScene
[>] StageManager              Storing previous frame pipe for SSRSpecular
[>] StageManager              Awaiting future pipe Exposure
[>] StageManager              Storing previous frame pipe for SMAAPostResolve
[>] UpdatePreviousPipesStage  Creating previous pipes stage ..
[>] StageManager              Writing shader config
[>] plugin:pssm               Initializing pssm ..
[>] ScatteringMethodEricBruneton Precomputing ...
[>] RenderPipeline            Finished initialization in 2.837 s
[>] RenderPipeline            Took 0.002 s until first frame
```

**Comment by tobspr** [comment]
Okay, great - I will just put the layout() qualifier before the uniform then - it seems it does not matter on nvidia (Hopefull it doesn't on AMD cards too).

Can you try if uniform restrict image3D xxx; works? (I.e., putting the restrict *before* the image3D).
Maybe its the same case here.

Hm, its hard to cell what causes the crash .. do you use a precompiled build of panda, or do you build yourself?

**Comment by germn** [comment]
> Can you try if uniform restrict image3D xxx; works?

Yes, you're right again: moving RESTRICT before type solves error.

> Hm, its hard to cell what causes the crash .. do you use a precompiled build of panda, or do you build yourself?

I'm using latest precompiled build from Panda site (Panda3D-SDK-1.10.0pre-380ff44.exe).

I just thought, that I got very similar behavior, when I tried to setup pipeline earlier. It happend because of "window-type offscreen" option (this bug reported [here](https://bugs.launchpad.net/panda3d/+bug/1495955)). I just removed this option everywhere.

May be now it crashes because of some Panda bug. I'll try to play with settings and code and will report if found something. Anyway, thank you for help!
